0000-interledger-protocol-4/0000-interledger-protocol-4.md

```diff
@@ -0,0 +1,146 @@
+7. Receiver checks the ILP Prepare packet, according to whatever is stipulated by the higher-level protocol (such as checking whether the amount received is above some minimum specified by the sender). 
+8. Receiver returns an ILP Fulfill packet, containing the preimage of the condition in the ILP Prepare packet. The receiver may include additional data, the formatting of which is determined by the higher-level protocol.
+9. Connector checks that the receiver has returned the fulfillment before the expiry in the ILP Prepare packet. If so, the connector returns the same ILP Fulfill packet to the previous connector and credits the receiver's account with the amount in the ILP Prepare packet. If the Prepare has expired, the connector returns an ILP Reject packet to the previous connector. Each connector repeats this step until the Fulfill (or Reject) packet reaches the sender.
+10. Sender checks that the preimage in the ILP Fulfill packet matches the condition in their Prepare packet and may read any data returned by the receiver.
+11. Sender signs an update to their payment channel with the first connector to cover the value of the ILP Prepare (or they may settle less frequently using slower on-ledger transfers or physical deliveries of cash or other goods) and communicates it to the connector.
+12. Connector validates the sender's signed payment channel update (or other proof of settlement) and increases the sender's bandwidth by the amount they paid. If the sender does not pay for the ILP Prepare, the connector may refuse to process further ILP packets until they do or the connector may block that user entirely.
+
+## Specification
+
+### Packet Format
+
+ILPv4 packets SHOULD be encoded using the ASN.1 Octet Encoding Rules.
+
+#### Type-Length Wrapper
+
+All ILP packets are wrapped in the following envelope:
+
+| Field | Type | Description |
+|---|---|---|
+| `type` | UInt8 | ID of the ILP Packet type |
+| `data` | Variable-Length Octet String | Packet contents, prefixed with their length |
+
+See the ASN.1 definition [here](../asn1/InterledgerPacket.asn).
+
+#### ILP Prepare
+
+ILP Prepare packets are `type` 12.
+
+| Field | Type | Description |
+|---|---|---|
+| `amount` | UInt64 | Local amount, denominated in the minimum divisible unit of the asset of the bilateral relationship. This field is modified by each connector, who applies their exchange rate before forwarding the packet |
+| `expiresAt` | Fixed-Length [Interledger Timestamp](../asn1/InterledgerTypes.asn) | Date and time when the packet expires. This field is modified by each connector, who subtracts some amount of time from the expiry before forwarding the packet |
+| `executionCondition` | UInt256 | Sha256 hash digest of the `fulfillment` that will execute the transfer of value represented by this packet |
+| `destination` | [ILP Address](../0015-ilp-addresses/0015-ilp-addresses.md) | ILP Address of the receiver |
```

Comment:
I would argue that connectors need one more piece of data that's not in the ILP packet: the sender from the previous hop.

The connector _can_ figure this out from the communication channel by which they received an ILP packet, but I suspect it would simplify implementations and make them far more flexible in terms of underlying communication layers if the packet contained a `from` field identifying the previous connector (or the originator in the first hop), probably also by ILP address.

With that information, a connector would know whose "bandwidth" to debit from just the ILP Packet itself. Although I guess you'd have to worry about spoofing other senders to tie up their "bandwidth"...

Reply:
Yeah, connectors should figure this out based on which communication channel they got the packet from. If there were a `from` field, it would just be one more thing the connector would need to verify because it would be easy to lie about it.

Reply:
That makes sense. You should find a place to state it explicitly in the standard.